doc/developer/design/20210412_postgres_sources.md

```diff
@@ -0,0 +1,210 @@
+2. **Decoding plugin**: An component that runs inside the upstream database that
+   is responsible for reading the raw WAL segments and reconstructing the
+   logical transaction messages before sending them over to the logical
+   replication receiver.
+3. **pgoutput**: The specific decoder that postgres uses when logical
+   replication is used. There are a lot of third-party plugins available but
+   pgoutput is offered by postgres itself.
+4. **PUBLICATION**: It is an object that tells the upstream database which
+   tables are to be published over a replication stream. This can be an explicit
+   list of table names or a special "all tables" publication that gets
+   everything.
+5. **replication slot**: It is an object that keeps track of how far along a
+   replica is with its stream consumption. Normally the upstream database has
+   one such object for each downstream replica. Each replica periodically sends
+   updates with the WAL offset up to which it has consumed which then lets the
+   upstream database safely clean up segments from disk.
+
+### Replication mechanism
+
+In order to replicate a set of tables from an upstream database a replica needs
+to orchestrate two phases, making sure that there is no data loss during the
+handover. The first phase involves the initial table sync which is done with
+normal `COPY` bulk transfer commands. Once the replica has the initial snapshot
+of a table it can switch to reading updates from a replication slot in order to
+keep its table synced.
+
+Ensuring that there is no gap between finishing the initial COPY and starting
+reading from the replication slot is done by synchronizing the database snapshot
+at which the `COPY` is performed with the starting point of the replication
+slot. The simplest way to do this is by using the following query structure:
+
+```sql
+-- Initial sync phase
+BEGIN READ ONLY ISOLATION LEVEL REPEATABLE READ
+CREATE_REPLICATION_SLOT <foo> LOGICAL pgoutput USE_SNAPSHOT;
+COPY `table` TO STDOUT;
+COMMIT
+
+-- Streaming phase
+START_REPLICATION SLOT <foo>
+```
+
+By creating the replication slot as the first statement in a transaction and
+setting the `USE_SNAPSHOT` option we ensure that the transaction and the slot
+are at the exact same point. The existence of the slot will prevent any WAL
+segments from being cleaned up while our replica is busy reading the initial
+snapshot.
+
+Once the initial sync is done the transaction is committed to release the
+snapshot and the replica can start streaming the slot normally with no data
+loss.
+
+### Transaction boundaries and timestamps
+
+A key design goal is to maintain the  transaction boundaries from the upstream
+database so that no query on materialize can observe states that never existed
+upstream. At the same time, the data ingested from postgres should be join-able
+with other sources in materialize. This means that all updates must have a
+timestamp that is related to the wall-clock time.
```

Comment:
Does it mean this? Many SQL systems have *system versioned time* and if they present that outwards it seems like it would be a very reasonable timestamp to use.

Reply:
What I mean is that currently in materialize there are BYO sources that have a timestamp value that could represent anything and realtime sources that are based on wall clock of ingestion.

You're right that two BYO sources could be using compatible system times and be perfectly join-able together. I'll re-write this sentence to make more clear

Reply:
Maybe a different framing from me then: It would be best if we mirrored the upstream systems view of times as faithfully as possible, and then in Materialize determined if we wanted to convert timestamps to local "materialize time", rather than baking in materialize time to the definition of the sources.